webrtc/README.md

```diff
@@ -0,0 +1,130 @@
+# WebRTC
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+|-----------------|---------------|--------|-----------------|
+| 1A              | Working Draft | Active |                 |
+
+Authors: [@mxinden]
+
+Interest Group: [@marten-seemann]
+
+## Motivation
+
+1. **No need for valid TLS certificates.** Enable browsers to connect to public
+   server nodes without those server nodes providing a TLS certificate within
+   the browsers trustchain. Note that we can not do this today with our
+   Websocket transport.
+
+2. **Hole punching in the browser**: Enable two browsers or a browser and a
+   non-public server node to connect.
+
+## Requirements
+
+- Loading a remote nodes certificate into ones browser trust-store is not an
+  option, i.e. doesn't scale.
+
+- No dependency on central STUN and/or TURN servers.
+
+## Protocol
+
+Scenario: Browser _A_ wants to connect to server node _B_.
+
+### Setup - Both _A_ and _B_
+
+1. [Generate a
+   certificate](https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-generatecertificate).
+
+2. [Get the certificate's
+   fingerprint](https://www.w3.org/TR/webrtc/#dom-rtccertificate-getfingerprints).
+
+3. Sign this fingerprint using the libp2p peer key to generate a "peer cert"
+
+### Connection Establishment
+
+1. Browser _A_ discovers server node _B_'s multiaddr, containing _B_'s IP, UDP
+  port and TLS certificate fingerprint (e.g.
+  `/ip6/2001:db8::/udp/1234/webrtc/<tls-certificate-fingerprint`), through some
+  external mechanism.
+
+2. _A_ create a connection with its local certificate.
+
+3. _A_ constructs _B_'s SDP offer based on _B_'s multiaddr and sets it
+   via
+   [`RTCPeerConnection.setRemoteDescription()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/setRemoteDescription).
+
+4. _A_ establishes the connection to _B_.
+
+5. _A_ opens a data stream.
+
+6. _A_ and _B_ exchange the "peer certs".
+
+7. [Get the remote peer's TLS
+   certificate](https://www.w3.org/TR/webrtc/#dom-rtcdtlstransport-getremotecertificates).
+
+8. Check the remote endpoint's "peer cert".
+
+## Open Questions
+
+_Where _Browser_ is replaced with the major desktop browsers of today (Chrome,
+Safari, Edge, Firefox)._
+
+### Direct connections
+
+#### Browser to public server node
+
+- Can _Browser_ generate a _valid_ SDP packet for the remote node based on the
+  remote's Multiaddr, where that Multiaddr contains the IP, UDP port and TLS
+  certificate fingerprint (e.g.
+  `/ip6/2001:db8::/udp/1234/webrtc/<tls-certificate-fingerprint`)? _Valid_ in
+  the sense that this generated SDP packet can then be used to establish a
+  WebRTC connection to the remote.
+
+- Do the major (Go / Rust / ...) WebRTC implementations allow us to accept a
```

Comment:
Go's pion does as well (https://github.com/pion/offline-browser-communication). I've gotten it to work non-offline (browser to remote server). Trying to understand why smoldot needs to add a certificate to the browser @melekes, because I've gotten the pion one to work remotely.

Reply:
cc: @wemeetagain

Reply:
> Trying to understand why smoldot needs to add a certificate to the browser @melekes

where this is coming from?

Reply:
From PR 2135 on smoldot. @wemeetagain asked a question there https://github.com/paritytech/smoldot/pull/2135/files#r848710012

Reply:
https://github.com/paritytech/smoldot/pull/2135#discussion_r877832337

Reply:
Resolving here. Feel free to re-open.